server/src/main/java/org/elasticsearch/action/admin/cluster/configuration/AddVotingConfigExclusionsRequest.java

```diff
@@ -56,32 +62,72 @@ public AddVotingConfigExclusionsRequest(String[] nodeDescriptions) {
      * @param timeout How long to wait for the added exclusions to take effect and be removed from the voting configuration.
      */
-    public AddVotingConfigExclusionsRequest(String[] nodeDescriptions, TimeValue timeout) {
+    public AddVotingConfigExclusionsRequest(String[] nodeDescriptions, String[] nodeIds, String[] nodeNames, TimeValue timeout) {
         if (timeout.compareTo(TimeValue.ZERO) < 0) {
             throw new IllegalArgumentException("timeout [" + timeout + "] must be non-negative");
         }
         this.nodeDescriptions = nodeDescriptions;
+        this.nodeIds = nodeIds;
+        this.nodeNames = nodeNames;
         this.timeout = timeout;
     }
 
     public AddVotingConfigExclusionsRequest(StreamInput in) throws IOException {
         super(in);
+        // TODO should this be removed in the latest version where nodeIds and nodeNames are used?
         nodeDescriptions = in.readStringArray();
+        // TODO which version to use here?
+        if (in.getVersion() == Version.V_EMPTY) {
+            nodeIds = in.readStringArray();
+            nodeNames = in.readStringArray();
+        }
         timeout = in.readTimeValue();
     }
 
     Set<VotingConfigExclusion> resolveVotingConfigExclusions(ClusterState currentState) {
         final DiscoveryNodes allNodes = currentState.nodes();
-        final Set<VotingConfigExclusion> resolvedNodes = Arrays.stream(allNodes.resolveNodes(nodeDescriptions))
+        Set<VotingConfigExclusion> allProcessedNodes = null;
+
+        if (nodeDescriptions.length >= 1) {
+            allProcessedNodes = Arrays.stream(allNodes.resolveNodes(nodeDescriptions)).map(allNodes::get)
+                                    .filter(DiscoveryNode::isMasterNode).map(VotingConfigExclusion::new).collect(Collectors.toSet());
+
+            if (allProcessedNodes.isEmpty()) {
+                throw new IllegalArgumentException("add voting config exclusions request for " + Arrays.asList(nodeDescriptions)
+                    + " matched no master-eligible nodes");
+            }
+        }
+        else {
+            DiscoveryNodes.NodeResolutionResults nodeResolutionResults;
+            Set<VotingConfigExclusion> resolvedNodes;
+            Set<VotingConfigExclusion> unresolvedNodes;
+
+            if (nodeIds.length >= 1) {
+                nodeResolutionResults = allNodes.resolveNodesExact(true, nodeIds);
+
+                unresolvedNodes = Arrays.stream(nodeResolutionResults.getUnresolvedNodes())
+                    .map(nodeId -> new VotingConfigExclusion(nodeId, "")).collect(Collectors.toSet());
+            }
+            else {
+                nodeResolutionResults = allNodes.resolveNodesExact(false, nodeNames);
+
+                unresolvedNodes = Arrays.stream(nodeResolutionResults.getUnresolvedNodes())
+                    .map(nodeName -> new VotingConfigExclusion("", nodeName)).collect(Collectors.toSet());
```

Comment:
Suggest using a sentinel value other than `""` here to clarify why the node name is missing. I suspect this will save some confusion in a future debugging session.

```suggestion
                    .map(nodeName -> new VotingConfigExclusion("_absent_", nodeName)).collect(Collectors.toSet());
```

Do these name-only exclusions do anything? IOW if you add an exclusion by name for an absent node, and then a node with that name joins the cluster, do we exclude it from the voting configuration? I think we don't today, but we should. I think it would also be helpful to update the `VotingConfigExclusion` to include the node  ID in that case.

Reply:
Changed to use `_absent_` here. 

>   I think we don't today, but we should. I think it would also be helpful to update the VotingConfigExclusion to include the node ID in that case.

Make sense, I'll push another commit for this. `VotingConfigExclusion` may need an additional flag to indicate its nodeId is set when node joins, so that later the set nodeId can be cleared when node leaves.

Reply:
I think it's ok to leave the node ID set when the node leaves. You shouldn't really have more than one node with the same name, at least not when you're using the node name to identify the node as a user would be doing here. If they want to go against recommendations and use duplicate node names then they'll need to specify node IDs in the API.

Reply:
I've pushed a new commit for this. Please let me know if that's the right place to put this logic.

Once these all look good, I'll start to work on adding more tests.

Reply:
Yes, the `JoinTaskExecutor` is indeed the right place to do this  :+1:

I think this means that no cluster states contain unresolved voting configurations which _could_ be resolved, which is an invariant that we should assert somewhere. Maybe in `Coordinator#improveConfiguration` since that's where we actually use the exclusions?

Reply:
> no cluster states contain unresolved voting configurations which could be resolved
In terms of code, do you mean that no node ID should be empty when it reaches `Coordinator#improveConfiguration` (as it needs to have value of either `_absent_` when no node with matching name joins yet, or an actual node  ID?) The reason I'm asking is currently there should not be any code path that set empty node ID, as we basically use `_absent_` to denote that,  so I'm not quite sure what needs to be asserted  there?

Reply:
I mean that every voting config exclusion with an ID of `_absent_` should not match any nodes currently in the cluster by name, and every voting config exclusion with a name of `_absent_` should not match any nodes currently in the cluster by ID (except, I guess, for nodes whose name is literally the string `"_absent_"`).

The point is that we are only resolving these things in the `JoinTaskExecutor` which today  is the only way that nodes can join the cluster, but we don't want to get caught out by any future changes that might generalise that.

Reply:
Ah I see, make sense. I've added the invariant check.